crates/ecs/src/lib.rs

```diff
@@ -641,323 +520,52 @@ pub struct World {
         &mut self,
         entity: Entity,
         component: ComponentType,
     ) -> WorldResult<()> {
-        let source_archetype_index = *self
-            .entity_to_archetype_index
-            .get(&entity)
-            .ok_or(WorldError::EntityDoesNotExist(entity))?;
-
-        let add = ArchetypeMutation::Addition(entity);
-
-        let (target_archetype_exists, source_type_ids, target_type_ids) =
-            self.find_target_archetype::<ComponentType>(add)?;
-
-        match target_archetype_exists {
-            Some(target_archetype_idx) => {
-                self.move_entity_components_between_archetypes(
-                    entity,
-                    target_archetype_idx,
-                    source_type_ids,
-                )?;
-
-                let target_archetype = self.archetypes
-                    .get_mut(target_archetype_idx)
-                    .expect("Archetype should exist since it was the requirement for entering this matching case");
-                target_archetype
-                    .add_component(component)
-                    .map_err(WorldError::CouldNotAddComponent)?;
-            }
-            None => {
-                let target_archetype_idx =
-                    self.move_entity_components_to_new_archetype(entity, source_type_ids)?;
-
-                let mut target_type_ids_vec: Vec<TypeId> = target_type_ids.into_iter().collect();
-
-                // Sort the vec since order of elements in vec matter when matching keys.
-                // Sorting the key is done both when inserting and reading from this hashmap to
-                // maintain consistent behaviour.
-                target_type_ids_vec.sort();
-                self.component_typeids_set_to_archetype_index
-                    .insert(target_type_ids_vec, target_archetype_idx);
-
-                let target_archetype = self
-                    .archetypes
-                    .get_mut(target_archetype_idx)
-                    .expect("This should be added in the previous function call");
-
-                // Handle incoming component
-                target_archetype.add_component_vec::<ComponentType>();
-                let archetype_indices = self
-                    .component_typeid_to_archetype_indices
-                    .entry(TypeId::of::<ComponentType>())
-                    .or_default();
-                archetype_indices.insert(target_archetype_idx);
-
-                target_archetype
-                    .add_component(component)
-                    .map_err(WorldError::CouldNotAddComponent)?;
-            }
-        }
-
-        let source_archetype = self
-            .archetypes
-            .get_mut(source_archetype_index)
-            .expect("Source archetype has already been fetched previously");
-
-        source_archetype.update_source_archetype_after_entity_move(entity);
-
-        Ok(())
+        self.add_component_to_entity_and_move_archetype(entity, component)
```

Comment:
Nit Do we want to rename `add_component_to_entity_and_move_archetype` to `add_component_type_to_entity` to follow the naming of `remove_component_type_from_entity`.